File: backend/app/models/product.py
from datetime import date, datetime
from typing import Optional

from sqlalchemy import Boolean, Date, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    category_id: Mapped[int] = mapped_column(ForeignKey("categories.id"), index=True)
    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str] = mapped_column(Text, default="")
    status: Mapped[str] = mapped_column(
        String(20), default="판매 예정", server_default="판매 예정"
    )
    harvest_start: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    harvest_end: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    sale_start: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    sale_end: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    is_deleted: Mapped[bool] = mapped_column(
        Boolean, default=False, server_default="false"
    )
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now()
    )

    category: Mapped["Category"] = relationship(back_populates="products")  # noqa: F821
    options: Mapped[list["ProductOption"]] = relationship(
        back_populates="product", cascade="all, delete-orphan"
    )
    images: Mapped[list["ProductImage"]] = relationship(
        back_populates="product", cascade="all, delete-orphan"
    )


class ProductOption(Base):
    __tablename__ = "product_options"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    product_id: Mapped[int] = mapped_column(
        ForeignKey("products.id", ondelete="CASCADE"), index=True
    )
    name: Mapped[str] = mapped_column(String(100))
    price: Mapped[int] = mapped_column(Integer)
    stock_quantity: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    stock_threshold: Mapped[int] = mapped_column(
        Integer, default=10, server_default="10"
    )
    sort_order: Mapped[int] = mapped_column(Integer, default=0, server_default="0")
    is_active: Mapped[bool] = mapped_column(
        Boolean, default=True, server_default="true"
    )

    product: Mapped["Product"] = relationship(back_populates="options")


class ProductImage(Base):
    __tablename__ = "product_images"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    product_id: Mapped[int] = mapped_column(
        ForeignKey("products.id", ondelete="CASCADE"), index=True
    )
    image_url: Mapped[str] = mapped_column(String(500))
    thumbnail_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    is_primary: Mapped[bool] = mapped_column(
        Boolean, default=False, server_default="false"
    )
    sort_order: Mapped[int] = mapped_column(Integer, default=0, server_default="0")

    product: Mapped["Product"] = relationship(back_populates="images")

File: admin/src/types/product.ts
export interface Category {
  id: number;
  name: string;
  sort_order: number;
}

export interface ProductOption {
  id?: number;
  name: string;
  price: number;
  stock_quantity: number;
  stock_threshold?: number;
  sort_order: number;
  is_active?: boolean;
}

export interface ProductImage {
  id?: number;
  image_url: string;
  thumbnail_url: string | null;
  is_primary?: boolean;
  sort_order: number;
}

export type ProductStatus = '판매 예정' | '판매 중' | '품절' | '판매 종료';

export interface Product {
  id: number;
  name: string;
  description: string;
  category_id: number;
  category?: Category;
  status: ProductStatus;
  harvest_start: string | null;
  harvest_end: string | null;
  sale_start: string | null;
  sale_end: string | null;
  options: ProductOption[];
  images: ProductImage[];
  created_at: string;
  updated_at: string;
}

export interface ProductListItem {
  id: number;
  name: string;
  category_id: number;
  category?: Category;
  status: ProductStatus;
  harvest_start: string | null;
  harvest_end: string | null;
  options: ProductOption[];
  primary_image: ProductImage | null;
  created_at: string;
}

export interface ProductImageCreate {
  image_url: string;
  thumbnail_url: string | null;
  sort_order: number;
}

export interface ProductCreate {
  name: string;
  description: string;
  category_id: number;
  status: ProductStatus;
  harvest_start?: string | null;
  harvest_end?: string | null;
  sale_start?: string | null;
  sale_end?: string | null;
  options: Omit<ProductOption, 'id'>[];
  images: ProductImageCreate[];
}

export interface ProductUpdate extends Partial<ProductCreate> {}

export interface ProductStatusUpdate {
  status: ProductStatus;
}
